### PR TITLE
[Design Discussion] Minimax

### PR DESCRIPTION
**Pull request by huiwang**
@tyrcho I tried with a new design which is a bit different from your negamax but I'm not sure it's a good one. My design guide is to separate problem representation from minimax algo as much as possible. As a result, I've taken the following decisions

1. Representation should distinguish active/terminal nodes. An active node should have at least one edge. A terminal node returns an empty edge list. 
2. Representation should handle player switch during branching(childOf()) operation. This is hidden from minimax algo.

What do you think of this design?

### CODE REVIEW

File: src/main/scala/com/truelaurel/codingame/minimax/MinimaxRepresentation.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package com.truelaurel.codingame.minimax
+
+/**
+  * Representation of a problem that can be solved with Minimax algorithm
+  *
+  * @tparam Node structure representing the game state
+  * @tparam Edge a legal action that can be applied on a node
+  */
+trait MinimaxRepresentation[Node, Edge] {
```
Comment (tyrcho) [inline src/main/scala/com/truelaurel/codingame/minimax/MinimaxRepresentation.scala]: This is actually very close to the trait GameRules, I'm not sure if naming it around the game concept or around the tree concept makes it clearer.

File: src/main/scala/com/truelaurel/codingame/minimax/MinimaxRepresentation.scala [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  /**
+    * Returns the child node by applying the edge to the given state.
+    */
+  def childOf(node: Node, edge: Edge): Node
+
+  /**
+    * Assess the given node. A Greater score means a better situation for the maximizing player
+    */
+  def assess(node: Node): Double
```
Comment (tyrcho) [inline src/main/scala/com/truelaurel/codingame/minimax/MinimaxRepresentation.scala]: this is the only method I pass separately in AlphaBetaAi (heuristic)

File: src/main/scala/com/truelaurel/codingame/minimax/MinimaxAlgorithm.scala [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+  type Repr = MinimaxRepresentation[Node, Edge]
+
+  def search(repr: Repr, depth: Int): Edge = {
+    val root = repr.root
+    val assessedEdges = for {
+      edge <- repr.edgesOf(root)
+      child = repr.childOf(root, edge)
+    } yield (edge, minimax(repr, child, depth - 1, maximizingPlayer = false))
+    assessedEdges.maxBy(_._2)._1
```
Comment (tyrcho) [inline src/main/scala/com/truelaurel/codingame/minimax/MinimaxAlgorithm.scala]: I know this is just a draft, but it makes code clearer to introduce a simple case class to name the elements rather than using the tuple accessors ._1 ._2 (see AlphaBetaAi.ScoredMove for a similar case)

File: src/main/scala/com/truelaurel/problem/tictactoe/minimax/TicTacToeMinimaxRepresentation.scala [hunk 4]
```diff
@@ -0,0 +1,61 @@
+package com.truelaurel.problem.tictactoe.minimax
+
+import com.truelaurel.codingame.minimax.MinimaxRepresentation
+
+/**
+  * Created by hwang on 18/06/2017.
+  */
+class TicTacToeMinimaxRepresentation(val root: TicTacToeState) extends MinimaxRepresentation[TicTacToeState, TicTacToeAction] {
```
Comment (tyrcho) [inline src/main/scala/com/truelaurel/problem/tictactoe/minimax/TicTacToeMinimaxRepresentation.scala]: GomokuBoard(3) is probably the same thing as this class

Comment (tyrcho) [inline src/main/scala/com/truelaurel/problem/tictactoe/minimax/TicTacToeMinimaxRepresentation.scala]: I'd like to be able to define the rules for a game in a class, and pick the algorithm I want to use with no impact on my game rules. 

Here it seems that we have to extend the parent type to define the game rules. @huiwang I prefer composition of inheritance too :)

File: src/main/scala/com/truelaurel/codingame/minimax/MinimaxRepresentation.scala (hunk 1 above)
Comment (huiwang) [inline src/main/scala/com/truelaurel/codingame/minimax/MinimaxRepresentation.scala]: This is indeed just a naming choice. The added value is not clear.

File: src/main/scala/com/truelaurel/codingame/minimax/MinimaxAlgorithm.scala (hunk 3 above)
Comment (huiwang) [inline src/main/scala/com/truelaurel/codingame/minimax/MinimaxAlgorithm.scala]: agreed for better readability

File: src/main/scala/com/truelaurel/problem/tictactoe/minimax/TicTacToeMinimaxRepresentation.scala (hunk 4 above)
Comment (huiwang) [inline src/main/scala/com/truelaurel/problem/tictactoe/minimax/TicTacToeMinimaxRepresentation.scala]: @tyrcho Definitely. We have to implement this class to be minimax-compliant. 

In terms of code reuse, I do find composition more appealing than inheritance. Here the context is more about defining the contract. Implementing this trait only means we are conform to the minmax algorithm.

Picking a minimax representation doesn't impose changes on the game rule. The game rule is always the same. A representation is about how we model the problem. A minimax repr modelizes a problem with a tree structure whereas a genetic repr modelizes a problem with genomes/populations. 

No mattter which representation is chosen, the game rule stays the same. The same rule to tell if a game is over. 

To demonstrate this idea, I'll try to refactor out the game rules into a separte class and solve the tic-tac-toe problem with GA by reusing the same rules but different representation.

**Comment by huiwang** [comment]
The suggested design aims to reduce as much as possible the set of constraints over a game's model. That being said, I think it would be possible to relax the two following contraints  
- nextPlayer method in GameState trait
- outcome method in GameRules trait 

Relaxing these two contrains means that 
- validMoves should return an empty set if the game is in a terminal state. 
- heuristic function should take into account game-over situations 
- the model should ensure that  game tree is expaned in a turn-based style

**Comment by tyrcho** [comment]
Sure ... I need to think a bit and it's too late for tonight :)

File: src/main/scala/com/truelaurel/problem/tictactoe/minimax/TicTacToeMinimaxRepresentation.scala (hunk 4 above)
Comment (tyrcho) [inline src/main/scala/com/truelaurel/problem/tictactoe/minimax/TicTacToeMinimaxRepresentation.scala]: OK, I think we more or less agree on the target.

Maybe we could have the following : 

traits to define the contract for 
- GameRules (specific for 2p)
- Problem with a tree structure
- Choosing next move (in 2p ? more ?) for a given GameRule is a problem with tree structure (still abstract)
- Problem with populations

Sample implementations : 
- Gomoku is a GameRule
- AlphaBeta/Minimax/negamax is an implementation to find the best element in a tree
- maybe MCTS also

Finding the best move in a 2p game is just combining the correct GameRule and implementation like AlphaBeta or MCTS.

This is almost the design I implemented with the package ai.scala.fp.game, but I missed the part about tree browsing which can be abstracted.

**Comment by tyrcho** [comment]
@huiwang I think the best way to "test" a new design is from the point of view of the developper writing code using our "framework". Would you share here maybe just this code without any real implementation yet ?

You could showcase :

- implementing GameRules for a trivial game like TicTacToe
- finding best move reusing an algo like MCTS (which does not require anything more than the rules)
- getting a Referee to get the steps and result of a game between 2 variants of the AI (like different time constraints)

Basically this would mean rewriting something similar to [GomokuDemo](https://github.com/TrueLaurel/CodinGame-Scala-Kit/blob/master/src/main/scala/ai/scala/fp/game/gomoku/GomokuDemo.scala) with the new API (without the implentations existing yet, just to discuss the design).

**Comment by tyrcho** [comment]
@Wei-1 , @trobert if you have the time, I'd like your opinion on this design discussion too ...

**Comment by huiwang** [comment]
@tyrcho 

I see more clearly why you separate game rules from heuristics. A game rule is objective. That's what defines a game. Heuristics are subjective. Different players may have different strategies to evaluate the game.

**Comment by huiwang** [comment]
Now regarding the nextPlayer constraint. 

By saying relax, I'm not saying that we can't use nextPlayer to tell who is to play for the next round. What I want to say is that we can leave the responsability on how to choose next player to the game rules. 

Regarding outcome, I do agree that it's important for a judge system to declare the winner when a game ends. However, from the point view of a minimax algorithm, it needs to know how the tree expands and how a tree node is evaluated. 

That's why I called that trait MinimaxRepresentation. It covers the minimum information that a game should provide in order to be solved by Minimax.

**Comment by huiwang** [comment]
I definitely agree that a client has the final word on the design quality. But I don't understand your question

> Would you share here maybe just this code without any real implementation yet ?

Do you want me to show the design with real impl or without? :)

**Comment by tyrcho** [comment]
@huiwang OK, I got it about the nextPlayer. You're right, this can be handled by the rules. On the other we have to encode somewhere in the GameState this information ... so why not expose it in the trait ? I'm not sure it makes a big difference to be honest.

> Do you want me to show the design with real impl or without? :)

*without* any impl ! it's enough to see the position of the client.

**Comment by huiwang** [comment]
For example, in the tic-tac-toe game, i don't need store a field named nextPlayer to check who is to play for the next round. I just need to compare the number of X and O to know the next player. 

This may sound stupid but it shows that it's a implementation detail that can be decided by the game.

**Comment by tyrcho** [comment]
@huiwang  you're right ... But I'm still not 100% convinced. ^^

From the client side, I think I prefer to call
`state.nextPlayer` rather than `rules.nextPlayer(state)` who will 99% of the time call `state.nextPlayer` in its impl.

It just feels awkward having a GameState which cannot tell who plays next.

Still "thinking aloud", maybe it means that GameState and GameRules cannot be easily decoupled.

Looking for references over the net ...

- https://stackoverflow.com/questions/361002/any-patterns-for-modelling-board-games
- https://en.wikipedia.org/wiki/Game_Description_Language

the [GDL spec](http://logic.stanford.edu/classes/cs227/2013/readings/gdl_spec.pdf) is probably a good starting point.

Still thinking (taking about 30 min to write this comment).

A positive point in removing this method from GameState trait is we can remove the trait entirely.

Also the same class can represent the same state for different game rules. Eg a grid can be used to play Tic Tac Toe (play until it's filled) or 3 men morris (play 3 stones then move them). Or a chess board can be used to play a chess variant. So the dependency is really from rules to board.

Hmm another point. When we ask a player (ai) its move, he needs to know which sides he plays (so the player ai can be stateless). It seems to me much easier if this information is encoded in the board state. `player.nextMove(gameState)`.

It might be just a gut feeling, but removing the nextPlayer member from GameState just worries me ...

**Comment by huiwang** [comment]
@tyrcho  

I'd like to make sure that we are well aligned on the design goal. 

For me, we are designing a contract between a problem and the minimax algo in order that the problem can be solved by the minimax. How to model a game is out of our design scope. We can model the game in GDL. Others can model the same game in other language.

Do you agree?

**Comment by tyrcho** [comment]
@huiwang well, my main goal was to design how to **find a good move in a turn based game**, and to provide generic implementations of MCTS / alphabeta for this purpose. For this I had in mind to specify 2 contracts : 

- what is a turn based game
- what is a ai which finds the next best move

Then the client can implement its game, select an ai (maybe he'll need to implement an extra heuristic function) and he gets a robot which can play the game.

It seems our objectives were not the same ...

**Comment by huiwang** [comment]
Consequently, it's not very astonishing that we can't converge. From the very beginning, my head was already filled with trees, nodes, edges.

But this is not a problem. I'm ready to stick with your current implementation (differences are tiny). Let's collect feedbacks from real-world contests and other developers using these algos. We'll improve the design based on these feedbacks.
